### ISSUE TITLE
False positive with `dataclasses.fields` inside `dataclass_transform`'d metaclass' `__call__`

### ISSUE THREAD

**Issue by patrick-kidger** [microsoft/pyright#4912]
```python
from dataclasses import fields
from typing import dataclass_transform

@dataclass_transform()
class Meta(type):
    def __call__(cls, *args, **kwargs):
        fields(cls)
```
produces
```
/myfile.py:7:16 - error: Argument of type "Self@Meta" cannot be assigned to parameter "class_or_instance" of type "DataclassInstance | Type[DataclassInstance]" in function "fields"
  Type "Self@Meta" cannot be assigned to type "DataclassInstance | Type[DataclassInstance]"
    "Meta" is incompatible with protocol "DataclassInstance"
      "__dataclass_fields__" is not present
    Type "Self@Meta" cannot be assigned to type "Type[DataclassInstance]" (reportGeneralTypeIssues)
```
Expected behaviour is no error.

Python version 3.8.16; pyright version 1.1.302 run from the command line

**Comment by patrick-kidger** [comment]
The same error also arises with:
```python
from dataclasses import fields
from typing import dataclass_transform

@dataclass_transform()
class Meta(type):
    pass

def f(m: Meta):
    fields(m)
```

**Comment by erictraut** [comment]
I think the current behavior of pyright is correct in this case. The type of the first parameter accepted by `fields` is `DataclassInstance | type[DataclassInstance]`, and `cls` is not compatible with this type. You'll need to find a workaround, such as `# type: ignore`.

**Comment by patrick-kidger** [comment]
How so? `cls` is an instance of `Meta`, and `dataclass_transform` declares that all instances of `Meta` are dataclasses?

**Comment by erictraut** [comment]
The `DataclassInstance` class is a protocol that requires that the class expose an attribute called `__dataclass_fields__`.

```python
class DataclassInstance(Protocol):
    __dataclass_fields__: ClassVar[dict[str, Field[Any]]]
```

The class `Meta` is a metaclass and does not expose such an attribute. This attribute appears only on a class that is created by the metaclass. The following type check as expected.

```python
class MyDataclass(metaclass=Meta):
    ...

x1: DataclassInstance = MyDataclass()
x2: type[DataclassInstance] = MyDataclass
```
